src/Transports.AspNetCore/WebSockets/WebSocketConnection.cs

```diff
@@ -0,0 +1,253 @@
+using System.Net.WebSockets;
+using GraphQL.Transport;
+using Microsoft.AspNetCore.Http;
+
+namespace GraphQL.Server.Transports.AspNetCore.WebSockets;
+
+/// <summary>
+/// Manages a WebSocket connection, dispatching messages to the specified <see cref="IOperationMessageProcessor"/>,
+/// and sending messages requested by the <see cref="IOperationMessageProcessor"/> implementation.
+/// <br/><br/>
+/// The <see cref="ExecuteAsync(IOperationMessageProcessor)"/> method may only be executed once for each
+/// instance. Awaiting the result will return once the WebSocket connection has been properly closed from both
+/// ends, after all messages have been sent.
+/// <br/><br/>
+/// Calls to <see cref="IOperationMessageProcessor.OnMessageReceivedAsync(OperationMessage)"/> be awaited
+/// before dispatching subsequent messages.
+/// <br/><br/>
+/// Calls to <see cref="CloseAsync()"/> and <see cref="SendMessageAsync(OperationMessage)"/> may be
+/// called on multiple threads simultaneously. They are queued for delivery and sent in the order posted.
+/// Messages posted after requesting the connection be closed will be discarded.
+/// </summary>
+public class WebSocketConnection : IWebSocketConnection
+{
+    private readonly WebSocket _webSocket;
+    private readonly AsyncMessagePump<Message> _pump;
+    private readonly IGraphQLSerializer _serializer;
+    private readonly WebSocketWriterStream _stream;
+    private readonly TaskCompletionSource<bool> _outputClosed = new();
+    private readonly int _closeTimeoutMs;
+    private int _executed;
+
+    /// <inheritdoc/>
+    public CancellationToken RequestAborted { get; }
+
+    /// <summary>
+    /// Returns the default disconnection timeout value.
+    /// See <see cref="GraphQLWebSocketOptions.DisconnectionTimeout"/>.
+    /// </summary>
+    protected virtual TimeSpan DefaultDisconnectionTimeout { get; } = TimeSpan.FromSeconds(10);
+
+    /// <inheritdoc/>
+    public DateTime LastMessageSentAt { get; private set; } = DateTime.UtcNow;
+
+    /// <inheritdoc/>
+    public HttpContext HttpContext { get; }
+
+    /// <summary>
+    /// Initializes an instance with the specified parameters.
+    /// </summary>
+    public WebSocketConnection(HttpContext httpContext, WebSocket webSocket, IGraphQLSerializer serializer, GraphQLHttpMiddlewareOptions options, CancellationToken cancellationToken)
```

Comment:
I would append a note that `cancellationToken` is for `RequestAborted` property or rename it into `requestAborted`.